### ISSUE TITLE
Genesis / Mega Drive Cheats Not Working as Intended

### ISSUE THREAD

**Issue by Draknarr** [libretro/RetroArch#4746]
Pardon my inexperience, I hope I am posting this correctly.  If not please let me know and I'll try to adjust accordingly.

The RetroArch community, I've found, has been having a lot of issues getting Cheats to work for RetroArch (particularly for RetroPie) for Genesis / Mega Drive.

I know there are multiple layers involved, and I'm still not sure how everything quite pieces together, but since there is a such a great out-of-the-box experience for Cheats for all the other console emulators, just not Genesis, I feel like *someone* would be interested in figuring out what's going on.

There are a lot of suggested work-arounds on the web, many of which I've tried, but to no avail (both for myself and others).  I'll try to list my workflow:
-I can easily Enter the Retroarch Menu -> Quick Menu -> Update Cheats, and get a list that works great for other consoles, and while I do have lists for Genesis, they don't do anything when applied.

-Potential Issue #1 - The cheats themselves are just wrong.  So I've gone through multiple sites, trying both Game Genie and Pro-Action Replay codes, and editing the .cht files.  It's such a simple process, so I was really hoping that was the solution.  And for some people it *does* work, but not everything (which is even weirder). 
---Some potential issues with codes being wrong (even though I suspect this is not the overall issue) -- Sometimes the code is missing two items.  So the default code will load as "FFFC3C:5F", but instead it needs to be "FFFC3C:005F", or "FFFC3C:5F5F", or something like that.  I also found, when looking at online sites, many of the codes are off by one digit. So for example, instead of being "FFEF51:0029", it's "FFEF50:0029".  Neither solution worked for most people, including myself, but seemed to help a few.  Again, why it helps some and not others I have no idea. 

-Potential Issue #2 - Looking for BIOS files?  I see some solutions (again, doesn't work for everyone) that they needed a ggenie.bin or proactionreplay file in their bios.  But it also needed some .cfg mods and got too complicated for me to follow. 

I'm testing on a Raspberry Pi 3 with RetroPie v4.1, which uses RetroArch.   In my searching I saw a similar thread on Reddit, and someone suggested the Original Poster post an Issue on GitHub, but I didn't see it so I thought I would.  

Again, thanks for the help.  This stuff fascinates me, and I'm happy to help test solutions, but my knowledge is quite limited =(.

**Comment by retro-wertz** [comment]
1. most cheats do not work - unless you can actually confirm that those cheats do work with actual game or other emulators
2. action replay seems to have different format, http://bsfree.shadowflareindustries.com/?s=11&d=5&g=1944

**Comment by Draknarr** [comment]
>Yes, so action replay does have a different format, and I've tried both versions.  But in both cases it's just inserting a .cht file in the correct format.  Where my knowledge falls short is if RetroArch only does one or the other (If it's a memory hack or ROM hack?), and if so if I need to dictate somewhere what type of hacks (game genie or action replay) it's trying to pull.   I can see my updated files, I can "apply" them, but I don't know what that's actually doing.

At least on a PC I know using something like Cheat Engine is finding specific memory locations and then freezing them to values, but I'm not sure exactly how the RetroArch is applying them, so it's harder for me to know exactly what's wrong.  Honestly I would be of the mindset that "no known game genie cheat would work on a different emulator," except for the fact that all the NES and SNES cheats pre-loaded work great.  

So I feel like there's got to be an easy solution somewhere within my grasp, I just don't know enough of how to grab it.

**Comment by retro-wertz** [comment]
retroarch has its own decoding(at least for game genie, i have not tested action replay format). no nid to select rom hack or memory has. just follow the examples and the formatting and it should work(assuming that the code does work though)

**Comment by Draknarr** [comment]
Oh interesting.  So it's not a case of anything being "broken", or any potential for there to be a "fix"? It's just that the codes are different for Genesis, and they fortunately happened to be the same for NES/SNES? 
I wish there was a way I could work out what the new values would be, but that would probably be a massive undertaking.

**Comment by fr500** [comment]
uhh...

as far as I know RetroArch doesn't do any kind of decoding of cheats... it just forwards the cheats to the core...

**Comment by Draknarr** [comment]
omg... fr500 .... I owe you a cookie!  I figured it out from your comment!!
I've been an idiot the whole time.  The secret was the core!
I'm going to write this up carefully so that future people can benefit from this...
I didn't realize the Core was the "type" of emulation you're running.  For Genesis, the RetroPie, by default, runs "picodrive".  picodrive must use different values for cheats.
So, the fix was simple... I just went to Retropie/configs/megadrive/emulators.cfg, and changed the line:
default = "lr-picodrive" to default = "lr-genesis-plus-gx", and then loaded the cheats as usual, and they worked!!!
Now, I should note, the cheat file I was using had a whole manner of stuff.  It had Pro Action Replay cheats, Game Genie Cheats, some with extra values, etc... so I don't know if the "default" cheats work or not.  BUT it means there's hope, and if the default cheats to NOT work, then it's just a matter of finding the correct cheat off a website.  

But thank you, that's the answer I needed, the answer was I just had to change the core it uses!
